atlasdb-client/src/main/java/com/palantir/atlasdb/cache/DefaultOffHeapCache.java

```diff
@@ -0,0 +1,178 @@
+        }
+        return value;
+    }
+
+    private Optional<V> getValue(K key) {
+        V inFlight = inFlightRequests.get(key);
+        if (inFlight != null) {
+            return Optional.of(inFlight);
+        }
+
+        return persistentStore.get(cacheDescriptor.get().storeNamespace(), key);
+    }
+
+    private static <K, V> CacheDescriptor createNamespaceAndConstructCacheProposal(
+            LogicalPersistentStore<K, V> store) {
+        StoreNamespace proposal = store.createNamespace(OFF_CACHE_NAMESPACE);
+        return ImmutableCacheDescriptor.builder()
+                .currentSize(new AtomicInteger())
+                .storeNamespace(proposal)
+                .build();
+    }
+
+    private static MetricName constructCacheMetricName(String metricSuffix) {
+        return MetricName.builder()
+                .safeName(MetricRegistry.name(DefaultOffHeapCache.class, metricSuffix))
+                .build();
+    }
+
+    private static class WriteBatcher <K, V>
+            implements CoalescingRequestFunction<Map.Entry<K, V>, Map.Entry<K, V>> {
+        DefaultOffHeapCache<K, V> offHeapCache;
+
+        WriteBatcher(DefaultOffHeapCache<K, V> offHeapCache) {
+            this.offHeapCache = offHeapCache;
+        }
+
+        @Override
+        public Map<Map.Entry<K, V>, Map.Entry<K, V>> apply(Set<Map.Entry<K, V>> request) {
+            CacheDescriptor cacheDescriptor = offHeapCache.cacheDescriptor.get();
+            if (cacheDescriptor.currentSize().get() >= offHeapCache.maxSize.getAsLong()) {
+                offHeapCache.taggedMetricRegistry.counter(CACHE_NUKE).inc();
+                offHeapCache.clear();
+            }
+            cacheDescriptor = offHeapCache.cacheDescriptor.get();
+            try {
+                Map<K, V> response = offHeapCache.persistentStore.get(
+                        cacheDescriptor.storeNamespace(),
+                        request.stream().map(Map.Entry::getKey).collect(Collectors.toList()));
+
+                int sizeIncrease = Sets.difference(request, response.entrySet()).size();
+                offHeapCache.persistentStore.put(
+                        cacheDescriptor.storeNamespace(),
+                        ImmutableMap.copyOf(request));
+
+                cacheDescriptor.currentSize().addAndGet(sizeIncrease);
+            } catch (SafeIllegalArgumentException exception) {
+                // happens when a store is dropped by a concurrent call to clear
+                log.warn("Clear called concurrently, writing failed", exception);
+            } finally {
+                offHeapCache.inFlightRequests.clear();
+            }
+            return KeyedStream.of(request.stream()).collectToMap();
+        }
+    }
+
+    @Value.Immutable
+    @Value.Style(visibility = Value.Style.ImplementationVisibility.PACKAGE)
+    interface CacheDescriptor {
+        AtomicInteger currentSize();
+        StoreNamespace storeNamespace();
```

Comment:
are we not changing store namespace? I'm really quite against it.

Reply:
No longer a `StoreNamespace` but a `StoreHandle`. I'm open to a different name.